### PR TITLE
Added UI Node

### PR DESCRIPTION
**Pull request by ChaoticTechSupport**
- Added a Base UI class
- Added a System Extend class for ease of use
- Modified Base UI class to contain Text Resizing options
- Added documentation.

### CODE REVIEW

**Comment by Yonodactyl** [comment]
For the sake of transparency, I don't know if we are ready to merge this in just yet.

I know it's nothing crazy, and it's really just setting some stuff up for once we get moving, but we are still hoping to have some standards established before we start working on code. I say we keep this PR opened, but it should hang around until we establish our processes.

**Comment by StrawberrySmoothieDev** [comment]
> For the sake of transparency, I don't know if we are ready to merge this in just yet.
> 
> I know it's nothing crazy, and it's really just setting some stuff up for once we get moving, but we are still hoping to have some standards established before we start working on code. I say we keep this PR opened, but it should hang around until we establish our processes.

Agreed. I'd say keep this in the request stage until we better figure out out baselines and foundation. For all we know we could be using 3D ui and this would just be bloat. I'll take a closer look at the code whenever I can to judge cleanliness but otherwise not much to do here.

File: Scripts/extend.gd
Comment (StrawberrySmoothieDev) [inline Scripts/extend.gd:1]: Name could cause issues. Furthermore, remember to use PascalCase for class names. Would recommend a class name that better describes what it does.

File: Scripts/extend.gd
Comment (StrawberrySmoothieDev) [inline Scripts/extend.gd:7]: If this is simply a func you want to be globally available, I would recommend making the script a global autoload singleton instead of a static func. Will cause way less issues. Furthermore, this function is (at least in my experience) not critical enough to be made global.

File: Scripts/ui/base/Ui.gd
Comment (StrawberrySmoothieDev) [inline Scripts/ui/base/Ui.gd:36]: Generally recommended to use pascal_case for variable definitions.

File: project.godot
Comment (StrawberrySmoothieDev) [inline project.godot:17]: Are you using godot .net? I am unsure if this will cause issues.

**Review by StrawberrySmoothieDev — CHANGES_REQUESTED** [review]
Overall pretty good. Take a look at the comments and resolve them when you can. I'll run a unit test and see if it actually works whenever I can. Thanks!

File: Scripts/extend.gd
Comment (ChaoticTechSupport) [reply Scripts/extend.gd:7]: actually I was going to have this as a sort of S.O.L class to add random methods that are commonly used, like books to integers for flipping,
or a change scene function. this was mainly an implementation for hard core programmers instead of developers, so adding it as a global class wouldn't be preforment or efficient as it would only stand to cause confusion. 

the reason that get all children should be in there is due to potential inventory, procedural, or basic implementation.

File: Scripts/ui/base/Ui.gd
Comment (ChaoticTechSupport) [reply Scripts/ui/base/Ui.gd:36]: ah, sorry I usually use PascalCase and forgot to switch to pascal_case in the for loop. I will fix that now.

File: project.godot
Comment (ChaoticTechSupport) [reply project.godot:17]: yes I am, as I usually develop in c# but I will switch to the regular hodot version now.

File: Scripts/extend.gd
Comment (StrawberrySmoothieDev) [reply Scripts/extend.gd:1]: Additionally the class name should probably be something along the lines of UtilityFunctions. Up to you.

File: Scripts/extend.gd
Comment (ChaoticTechSupport) [reply Scripts/extend.gd:7]: utility Functions does sound better than Extend

**Review by StrawberrySmoothieDev — COMMENTED** [review]
Looks great! Keeping this on the table for now, but it's got my seal of approval. Waiting for review from community members and @GodotCommunityGames

**Comment by KaizNike** [comment]
Not robust enough for my liking. Works well on a single Label, resizes well on Y axis, some cutoff on X. When I tried a more complex UI structure it didn't do anything for it due to the fact it requires each and every control node's font to be manually resized. Where most times I set the font size on a parent node and let that dictate what size the children are, but it's close to greatness!

**Comment by ChaoticTechSupport** [comment]
> Not robust enough for my liking. Works well on a single Label, resizes well on Y axis, some cutoff on X. When I tried a more complex UI structure it didn't do anything for it due to the fact it requires each and every control node's font to be manually resized. Where most times I set the font size on a parent node and let that dictate what size the children are, but it's close to greatness!

did you try the other modes? also, do yo mean you use theme and not theme override? because I can actually activate theme mode in the code if that's what your talking about.

**Comment by KaizNike** [comment]
Yeah I meant theme support for this text resize UX. I got it working for a moment on my own, but its prone to errors. For example if for any reason there is a non-Control child it crashes. 

Here are my changes to UserInterface.gd to review:
`func _ready() -> void:
	# Store font size and node
	for node: Node in UtilityFunctions.get_all_Children(self):
		if not node.is_class("Control"):	#Safety
			continue
		if node.has_theme_font_size_override("font_size"):
			font_nodes[node] = node.get("theme_override_font_sizes/font_size")
		elif self.theme is Theme:
			if self.theme.has_default_font_size() and node.get_theme_default_font_size():	#Addition to cover themes
				font_nodes[node] = node.get_theme_default_font_size()
	
	get_tree().get_root().size_changed.connect(font_resize)
	font_resize()`

What do you think?

**Comment by ChaoticTechSupport** [comment]
> Yeah I meant theme support for this text resize UX. I got it working for a moment on my own, but its prone to errors. For example if for any reason there is a non-Control child it crashes.
> 
> Here are my changes to UserInterface.gd to review:
> 
> ```
> func _ready() -> void:
> 	# Store font size and node
> 	for node: Node in UtilityFunctions.get_all_Children(self):
> 		if not node.is_class("Control"):	#Safety
> 			continue
> 		if node.has_theme_font_size_override("font_size"):
> 			font_nodes[node] = node.get("theme_override_font_sizes/font_size")
> 		elif self.theme is Theme:
> 			if self.theme.has_default_font_size() and node.get_theme_default_font_size():	#Addition to cover themes
> 				font_nodes[node] = node.get_theme_default_font_size()
> 	
> 	get_tree().get_root().size_changed.connect(font_resize)
> 	font_resize()
> ```
> 
> What do you think?

i just got back and finished in 5 minutes subbmitting for review now.

**Comment by KaizNike** [comment]
I think it's looking perfectly up to par now.

**Comment by ChaoticTechSupport** [comment]
> I think it's looking perfectly up to par now.
> 
> EDIT: Noticed the focus start doesn't work as expected for TabContainers (for those that use em like me) as the focus is set to an internal node.

thanks for telling me! I'll look into that and hopefully have it fixed soon.

**Comment by ChaoticTechSupport** [comment]
> I think it's looking perfectly up to par now.
> 
> EDIT: Noticed the focus start doesn't work as expected for TabContainers (for those that use em like me) as the focus is set to an internal node.

hi! it is now fixed! if you want to take a look see if somthings off that would be great!

**Review by KaizNike — APPROVED** [review]
Works great on my end.

**Comment by StrawberrySmoothieDev** [comment]
Taking a look for a approval. Think this one is general enough to be approved even now.

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:14]: Generally put enum members in MACRO_CASE.

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:21]: Remember to document your variables. Stuff like this is fairly self-explanatory but the ones below could use with some docs.

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:32]: MACRO_CASE

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:26]: Documentation for vars should be inline with the variable declaration.

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:67]: Function documentation should be inline w/function declaration.

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:88]: Try breaking up these long conditionals with newlines. GDscript offers this via backslashes.
 `	if owner \
	and get_child(0):
		pass`

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UI/Base/UserInterface.gd:103]: This is good for internal debug, but push_error is better for final releases. Does not need a change now, but useful to consider.

File: Scripts/UtilityFunctions.gd
Comment (StrawberrySmoothieDev) [inline Scripts/UtilityFunctions.gd:7]: This implementation is fine for now, but later down the line we may want to replace/refactor it. No changes needed atm.

**Review by StrawberrySmoothieDev — CHANGES_REQUESTED** [review]
Looks pretty good, made some suggestions, however, with some testing I might recommend making it work not just on viewport resize but also on the resizing of the container/parent control node. Also is there any reason for making it work via a parent and not directly in the label? Thanks! :3

**Comment by ChaoticTechSupport** [comment]
> Looks pretty good, made some suggestions, however, with some testing I might recommend making it work not just on viewport resize but also on the resizing of the container/parent control node. Also is there any reason for making it work via a parent and not directly in the label? Thanks! :3

Hi Smoothie, this setup is meant to ensure the game looks consistent on any monitor. It's not designed to actively resize containers. If you're looking for text resizing within containers, that's a different feature from what I've implemented. Resizing containers instead of adjusting the viewport wouldn't make sense, as both methods would give the same result. However, container resizing would be less effective, as it could break animations and take up more storage space.

if you have a specific idea which you would want container based text resizing then i would implament it as a seperate feature but otherwise i dont see the use.

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd:14]: Fixed

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd:21]: sorry i was rushing through focus and forgot to add docs

**Comment by KaizNike** [comment]
> > Looks pretty good, made some suggestions, however, with some testing I might recommend making it work not just on viewport resize but also on the resizing of the container/parent control node. Also is there any reason for making it work via a parent and not directly in the label? Thanks! :3
> 
> Hi Smoothie, this setup is meant to ensure the game looks consistent on any monitor. It's not designed to actively resize containers. If you're looking for text resizing within containers, that's a different feature from what I've implemented. Resizing containers instead of adjusting the viewport wouldn't make sense, as both methods would give the same result. However, container resizing would be less effective, as it could break animations and take up more storage space.
> 
> if you have a specific idea which you would want container based text resizing then i would implament it as a seperate feature but otherwise i dont see the use.

I agree, container text resizing isn't in the scope of this rather small but useful script.

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd:67]: i would argue that this is a bad idea for the same reason that you should not make a class documentation inline with the class name, but i will fix it.

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd:88]: ah, i was looking for that but couldnt find it, thanks.

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd:103]: i was also looking for that function, but could not find it, thank you.

File: Scripts/UtilityFunctions.gd
Comment (ChaoticTechSupport) [reply Scripts/UtilityFunctions.gd:7]: do you mean becuse it is recursive?

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd:32]: fixed

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd]: fixed

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd]: fixed

File: Scripts/UI/Base/UserInterface.gd
Comment (ChaoticTechSupport) [reply Scripts/UI/Base/UserInterface.gd]: fixed.

File: Scripts/UI/Base/UserInterface.gd
Comment (StrawberrySmoothieDev) [reply Scripts/UI/Base/UserInterface.gd]: Otherwise it won't show up in the autogenerated documentation. I agree it's a bit scuffed, but just how it is atm. Might make an issue in the godot repo.

File: Scripts/UtilityFunctions.gd
Comment (StrawberrySmoothieDev) [reply Scripts/UtilityFunctions.gd]: More so merging it with a community global functions script.

**Comment by StrawberrySmoothieDev** [comment]
> > Looks pretty good, made some suggestions, however, with some testing I might recommend making it work not just on viewport resize but also on the resizing of the container/parent control node. Also is there any reason for making it work via a parent and not directly in the label? Thanks! :3
> 
> Hi Smoothie, this setup is meant to ensure the game looks consistent on any monitor. It's not designed to actively resize containers. If you're looking for text resizing within containers, that's a different feature from what I've implemented. Resizing containers instead of adjusting the viewport wouldn't make sense, as both methods would give the same result. However, container resizing would be less effective, as it could break animations and take up more storage space.
> 
> if you have a specific idea which you would want container based text resizing then i would implament it as a seperate feature but otherwise i dont see the use.
Noted. Looks good to me :3

**Review by StrawberrySmoothieDev — APPROVED** [review]
Looks great! Well done :3 I really appreciate it, sorry if I'm a bit tight on standards but I like to run a tight ship.

**Pull request merged**
